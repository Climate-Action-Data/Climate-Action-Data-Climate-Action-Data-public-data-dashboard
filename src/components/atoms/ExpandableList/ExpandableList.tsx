import { generateRandomString } from '@/utils/GenerationHelpers'
import { Box, Button, Hide, Text } from '@chakra-ui/react'
import { useState } from 'react'
import { useTranslation } from 'react-i18next'

interface ExpandableListProps {
  items: string[]
  itemLimit?: number
}

const DEFAULT_ITEM_LIMIT = 4

export const ExpandableList = (props: ExpandableListProps) => {
  const actualProps = {
    ...props,
    itemLimit: props?.itemLimit ?? DEFAULT_ITEM_LIMIT,
  }
  const { items, itemLimit } = actualProps
  const [isExpanded, setIsExpanded] = useState(false)
  const { t } = useTranslation(`projectDetails`)
  const { t: tHome } = useTranslation(`home`)

  if (items.length === 0) {
    return <Text>{tHome(`noData`)}</Text>
  }

  const elementToRender = isExpanded ? items.length : itemLimit

  const renderItemList = () => {
    return items
      .map((item, index) => {
        return <Text key={`expandable-${generateRandomString()}`}>{item ?? tHome(`noData`)}</Text>
      })
      .slice(0, elementToRender)
  }

  const handleExpand = () => {
    setIsExpanded(!isExpanded)
  }

  return (
    <Box>
      {renderItemList()}
      {items.length > itemLimit && (
        <Hide below="md">
          <Button data-testid="verification-expand" onClick={handleExpand} variant="underlinedLink" style={{ color: `#00242C`, fontWeight: 500 }}>
            {isExpanded ? t(`viewLess`) : t(`viewMore`)}
          </Button>
        </Hide>
      )}
    </Box>
  )
}
