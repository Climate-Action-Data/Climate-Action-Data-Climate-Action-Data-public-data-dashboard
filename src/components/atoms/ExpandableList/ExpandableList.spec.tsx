import { fireEvent, render, screen } from '@testing-library/react'
import { ExpandableList } from './ExpandableList'

const ITEMS = [`item1`, `item2`, `item3`, `item4`, `item5`]

describe(`ProjectDetailsVerificationList`, () => {
  it(`renders correctly with full array`, () => {
    const { container } = render(<ExpandableList items={ITEMS} />)
    expect(container).toMatchSnapshot()
  })

  it(`renders correctly with full array and limit`, () => {
    const { container } = render(<ExpandableList items={ITEMS} itemLimit={2} />)
    expect(container).toMatchSnapshot()
  })

  it(`renders correctly with full array and click`, () => {
    const { container } = render(<ExpandableList items={ITEMS} />)
    const expandButton = screen.getByTestId(`verification-expand`)
    fireEvent.click(expandButton)
    expect(container).toMatchSnapshot()
  })

  it(`renders correctly with empty array`, () => {
    const { container } = render(<ExpandableList items={[]} />)
    expect(container).toMatchSnapshot()
  })
})
