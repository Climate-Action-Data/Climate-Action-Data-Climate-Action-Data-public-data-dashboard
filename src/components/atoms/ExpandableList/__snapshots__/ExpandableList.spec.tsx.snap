// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetailsVerificationList renders correctly with empty array 1`] = `
<div>
  <p
    class="chakra-text css-0"
  >
    noData
  </p>
</div>
`;

exports[`ProjectDetailsVerificationList renders correctly with full array 1`] = `
<div>
  <div
    class="css-0"
  >
    <p
      class="chakra-text css-0"
    >
      item1
    </p>
    <p
      class="chakra-text css-0"
    >
      item2
    </p>
    <p
      class="chakra-text css-0"
    >
      item3
    </p>
    <p
      class="chakra-text css-0"
    >
      item4
    </p>
    <button
      class="chakra-button css-4xx2wk"
      data-testid="verification-expand"
      type="button"
    >
      View More
    </button>
  </div>
</div>
`;

exports[`ProjectDetailsVerificationList renders correctly with full array and click 1`] = `
<div>
  <div
    class="css-0"
  >
    <p
      class="chakra-text css-0"
    >
      item1
    </p>
    <p
      class="chakra-text css-0"
    >
      item2
    </p>
    <p
      class="chakra-text css-0"
    >
      item3
    </p>
    <p
      class="chakra-text css-0"
    >
      item4
    </p>
    <p
      class="chakra-text css-0"
    >
      item5
    </p>
    <button
      class="chakra-button css-4xx2wk"
      data-testid="verification-expand"
      type="button"
    >
      View Less
    </button>
  </div>
</div>
`;

exports[`ProjectDetailsVerificationList renders correctly with full array and limit 1`] = `
<div>
  <div
    class="css-0"
  >
    <p
      class="chakra-text css-0"
    >
      item1
    </p>
    <p
      class="chakra-text css-0"
    >
      item2
    </p>
    <button
      class="chakra-button css-4xx2wk"
      data-testid="verification-expand"
      type="button"
    >
      View More
    </button>
  </div>
</div>
`;
